### PR TITLE
Topo service

### PR DESCRIPTION
**Pull request by ybc1991**

### CODE REVIEW

File: bluefog/torch/topology_util.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    StarGraph,
+    SymmetricExponentialGraph,
+)
+
+from bluefog.common.topology_util import (
+    GetDynamicOnePeerSendRecvRanks,
+    GetExp2DynamicSendRecvMachineRanks,
+    GetInnerOuterRingDynamicSendRecvRanks,
+    GetInnerOuterExpo2DynamicSendRecvRanks,
+)
```
Comment (hanbinhu) [inline bluefog/torch/topology_util.py:30]: Why importing everything here?

File: test/torch_basics_test.py [hunk 2]
```diff
@@ -113,37 +128,64 @@ def test_in_out_neighbors_expo2(self):
...
+    src_ranks = infer_destination_source_ranks(dst_ranks=out_neighbors)
+    assert sorted(src_ranks) == in_neighbors
+    dst_ranks = infer_destination_source_ranks(src_ranks=in_neighbors)
+    assert sorted(dst_ranks) == out_neighbors
+
+    W = infer_destination_source_ranks(
+        dst_ranks=out_neighbors, construct_adjacency_matrix=True
+    )
+    expected_W = (nx.to_numpy_array(topo) > 0).astype(float)
+    expected_W /= expected_W.sum(axis=0)
```
Comment (hanbinhu) [inline test/torch_basics_test.py:182]: Move expected_W computation above `infer_destination_souce_ranks` with True `construct_adjacency_matrix`

File: test/torch_basics_test.py [hunk 3]
```diff
@@ -113,37 +128,64 @@ def test_in_out_neighbors_expo2(self):
...
+    out_neighbors = bf.out_neighbor_ranks()
+
+    src_ranks = infer_destination_source_ranks(dst_ranks=out_neighbors)
+    assert sorted(src_ranks) == in_neighbors
+    dst_ranks = infer_destination_source_ranks(src_ranks=in_neighbors)
+    assert sorted(dst_ranks) == out_neighbors
+
+    W = infer_destination_source_ranks(
+        dst_ranks=out_neighbors, construct_adjacency_matrix=True
+    )
```
Comment (hanbinhu) [inline test/torch_basics_test.py:180]: Also return src_ranks?

File: test/torch_basics_test.py [hunk 4]
```diff
@@ -113,37 +128,64 @@ def test_in_out_neighbors_expo2(self):
...
+    expected_W = (nx.to_numpy_array(topo) > 0).astype(float)
+    expected_W /= expected_W.sum(axis=0)
+    np.testing.assert_allclose(W, expected_W)
+    W = infer_destination_source_ranks(
+        src_ranks=in_neighbors, construct_adjacency_matrix=True
+    )
```
Comment (hanbinhu) [inline test/torch_basics_test.py:186]: Also return dst_ranks?

File: bluefog/torch/topology_util.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        dst_ranks: A list of destination ranks. If provided the dst_link, a corresponding
+            src_ranks will be returned.
+        src_ranks: A list of destination ranks. If provided the src_ranks, a corresponding
+            dst_link will be returned.
+        construct_adjacency_matrix: If true, adjacency matrix will be return instead.
+            Element w_{ij} represents the weights sending from node i to node j. 
+            We use column normalized style, i.e. the sum of receiving weight is 1.
+
+    Raises:
+        ValueError: If dst_ranks or src_ranks does not contain integer from 0 to size-1.
```
Comment (hanbinhu) [inline bluefog/torch/topology_util.py:64]: Mention dst_ranks and src_ranks should not be given at the same time

**Review by hanbinhu — CHANGES_REQUESTED** [review]

File: test/torch_basics_test.py (hunk 4 above)
Comment (ybc1991) [reply test/torch_basics_test.py:184]: done

File: test/torch_basics_test.py (hunk 3 above)
Comment (ybc1991) [reply test/torch_basics_test.py:182]: done

File: bluefog/torch/topology_util.py (hunk 5 above)
Comment (ybc1991) [reply bluefog/torch/topology_util.py:34]: Split the function into two

File: test/torch_basics_test.py (hunk 2 above)
Comment (ybc1991) [reply test/torch_basics_test.py:201]: Done

File: bluefog/torch/topology_util.py (hunk 1 above)
Comment (ybc1991) [reply bluefog/torch/topology_util.py]: Removed. Used to allow it to import by bluefog.torch.topology_util. (Not good)

File: examples/construct_topo.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+elif bf.rank() == 1:
+    dst_list = [0]
+elif bf.rank() == 3:
+    dst_list = [0]
+else:
+    dst_list = [0]
+
+bf.set_topology(bf.topology_util.ExponentialTwoGraph(size=4))
+print(f"{bf.rank()}: {dst_list}")
+# print(bf.rank() in dst_list)
```
Comment (hanbinhu) [inline examples/construct_topo.py:20]: Delete all unnecessary comments in this file.

File: examples/construct_topo.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+if bf.rank() == 0:
+    dst_list = [1, 2]
+elif bf.rank() == 1:
+    dst_list = [0]
+elif bf.rank() == 3:
+    dst_list = [0]
+else:
+    dst_list = [0]
+
+bf.set_topology(bf.topology_util.ExponentialTwoGraph(size=4))
```
Comment (hanbinhu) [inline examples/construct_topo.py:18]: Add an assertion for bf.size()==4

**Review by hanbinhu — CHANGES_REQUESTED** [review]

File: examples/construct_topo.py (hunk 6 above)
Comment (ybc1991) [reply examples/construct_topo.py:20]: that file should not be checked in....

File: examples/construct_topo.py (hunk 7 above)
Comment (ybc1991) [reply examples/construct_topo.py:18]: that file should not be checked in....

**Review by hanbinhu — APPROVED** [review]
LGTM

**Pull request merged**